ptbcontrib/postgres_persistence/postgrespersistence.py

```diff
@@ -0,0 +1,142 @@
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser Public License for more details.
+#
+# You should have received a copy of the GNU Lesser Public License
+# along with this program.  If not, see [http://www.gnu.org/licenses/].
+"""This module contains PostgresqlPersistence class"""
+
+
+from logging import getLogger
+from collections import defaultdict
+from typing import Dict, Any, Callable
+
+from telegram.ext import DictPersistence
+from telegram.utils.helpers import (
+    encode_conversations_to_json,
+    decode_conversations_from_json,
+)
+
+from sqlalchemy.orm import scoped_session
+from sqlalchemy.sql import text
+
+import ujson as json
+
+
+class PostgresPersistence(DictPersistence):
+    """Using Postgresql database to make user/chat/bot data persistent across reboots.
+
+    Attributes:
+        store_user_data (:obj:`bool`): Whether user_data should be saved by this
+            persistence class.
+        store_chat_data (:obj:`bool`): Whether chat_data should be saved by this
+            persistence class.
+        store_bot_data (:obj:`bool`): Whether bot_data should be saved by this
+            persistence class.
+
+    Args:
+        session (:obj:`scoped_session`, Required): sqlalchemy scoped session.
+        store_user_data (:obj:`bool`, optional): Whether user_data should be saved by this
+            persistence class. Default is :obj:`True`.
+        store_chat_data (:obj:`bool`, optional): Whether user_data should be saved by this
+            persistence class. Default is :obj:`True`.
+        store_bot_data (:obj:`bool`, optional): Whether bot_data should be saved by this
+            persistence class. Default is :obj:`True` .
+
+    """
+
+    def __init__(
+        self,
+        session: scoped_session,
+        store_user_data: bool = True,
+        store_chat_data: bool = True,
+        store_bot_data: bool = True,
+    ) -> None:
+
+        super().__init__(
+            store_user_data=store_user_data,
+            store_chat_data=store_chat_data,
+            store_bot_data=store_bot_data,
+        )
+
+        self.logger = getLogger(__name__)
+        if not isinstance(session, scoped_session):
+            raise TypeError("session must be `sqlalchemy.orm.scoped_session` object")
+
+        self._session = session
+        self.__init_database()
+        self.__load_database()
+
+    def __init_database(self) -> None:
+        """creates table for persistence data if not exists"""
+
+        create_table_qry = """
+            CREATE TABLE IF NOT EXISTS persistence(
+            data json NOT NULL);"""
+        self._session.execute(text(create_table_qry))
+        self._session.commit()
+
+    def __load_database(self) -> None:
```

Comment:
Didn't try to understand in detail what's happening in here, so again just a thought: `DictPersistence` can accept json-string to recover the data from on init. Would loading those strings from DB and passing it to the parents `__init__` work?

Reply:
I still need to divide it into separate json strings, i.e for `user/chat bot_data` and conversation data, as json is saved into database as a single json string by using single row only without any primary key.. so ig it'll be better if we leave it as it is.

Reply:
Adding primary key setup will be mess as postgresql is rational database so we'll need to make 4 different columns then save then using 4 unique primary keys

Reply:
> I still need to divide it into separate json strings, i.e for `user/chat bot_data` and conversation data, as json is saved into database as a single json string by using single row only, without any primary key.. so ig it'll be better if we leave it as it is.

yeah. okay. was just a thought ;)